### PR TITLE
Make manager state consistent and allow to synchronize in the handler

### PR DESCRIPTION
**Pull request by emcastillo**
we should load the state with `exeuction=iteration` since the on-going executions do not exists when restoring the state, and this leads to an inconsistent state.

cc @HiroakiMikami

### CODE REVIEW

**Comment by kmaehashi** [comment]
* Could you update tests as well?
* States of extensions with `is_async = True` flag also needs this.

**Comment by emcastillo** [comment]
PTAL

File: pytorch_pfn_extras/handler.py [hunk 1]
```diff
@@ -205,6 +206,9 @@ def __init__(self, logic, entry_runtime, options):
                     Default is an empty list.
                 * ``'async'`` (bool):
                     If ``True``, async mode is enabled. Default is ``False``.
+                * ``'sync_trigger'`` (tuple or Trigger):
+                    Trigger used to synchronize iterations in
+                    asynchronous mode. Default is ``None``.
```
Comment (kmaehashi) [inline pytorch_pfn_extras/handler.py:211]: Do we need to pass the trigger manually? I was thinking that `_synchronize_train` is automatically called for all `needs_model_state = True` extensions.

**Comment by emcastillo** [comment]
I did an alternative implementation where I add a `needs_sync` attr to the extensions and the target object to be synced (manager does not have direct access to trainer/handler), This needs to increase the public APIs of handler and trainer.

Personally I think that passing the trigger as a handler option is much cleaner 😅

**Pull request closed without merging**